doc/adr/0003-nats-headers.md

```diff
@@ -0,0 +1,72 @@
+# 3. nats-headers
+
+Date: 2020-05-28
+
+## Status
+
+Accepted
+
+## Context
+
+This document describes NATS Headers from the perspective of clients. NATS headers allow clients to specify additional metadata in the form of headers. The headers borrow strongly from [HTTP Headers](https://tools.ietf.org/html/rfc7230#section-3.2).
+
+The salient points of the HTTP header specification are:
+
+- Each header field consists of a case-insensitive field name followed by a colon (":"), optional leading whitespace, the field value, and optional trailing whitespace.
+- No spaces are allowed between the header field name and colon.
+- Field value may be preceeded or followed by optional whitespace.
+- The specification may allow any number of strange things like comments/tokens etc.
+- The keys can repeat.
```

Comment:
At the protocol level, internally though the `http.Header` lib will dedup the keys since represented as a map so header may look different on another hop. Also maybe should note that keys aren't in any particular order, even though the `http.Header` library will sort them 'alphabetically' when publishing them through the wire.

Reply:
```
// Add adds the key, value pair to the header.
// It appends to any existing values associated with key.
// The key is case insensitive; it is canonicalized by
// CanonicalHeaderKey.
func (h Header) Add(key, value string) {
	textproto.MIMEHeader(h).Add(key, value)
}

// Values returns all values associated with the given key.
// It is case insensitive; textproto.CanonicalMIMEHeaderKey is
// used to canonicalize the provided key. To use non-canonical
// keys, access the map directly.
// The returned slice is not a copy.
func (h Header) Values(key string) []string {
	return textproto.MIMEHeader(h).Values(key)
}
```

Reply:
https://github.com/golang/go/blob/master/src/net/http/header.go#L21
https://github.com/golang/go/blob/master/src/net/http/header.go#L154

Reply:
anyway, just to comment that although the http header Go lib does a sort of the keys, for the nats http header spec similar to the http rfc we don't really define an order right now.

```go
m := nats.NewMsg(subject)
m.Header.Add("foo", "1")
m.Header.Add("bar", "2")
m.Header.Add("quux", "3")
m.Header.Add("baz", "4")
nc.PublishMsg(m)

// Result:
// NATS/1.0
// Bar: 2
// Baz: 4
// Foo: 1
// Quux: 3
```